### PR TITLE
Add DefaultValue.disallow

### LINKED ISSUE #1541: Reading from an `Event` trait inside a `Tuple` trait should be an error
Reading the value of an `Event()` trait should be an error, even when it's contained in another trait type. But we have the following behaviour in current Traits:
```python
>>> from traits.api import HasTraits, Event, Tuple
>>> 
>>> class A(HasTraits):
...     foo = Tuple(Event(), Event())
... 
>>> a = A()
>>> a.foo
(<undefined>, <undefined>)
```

### PR DESCRIPTION
**Pull request by mdickinson**
This PR adds a new value to the `DefaultValue` enumeration: `DefaultValue.disallow`. The intent is to use this for trait types that don't have a valid default. Right now that's `Event` and `Disallow`, but it could potentially be extended to other trait types in the future. The main feature of `DefaultValue.disallow` is that an attempt to get the default value of the corresponding CTrait instance using `default_value_for` will fail, raising `ValueError`.

Rationale: With the current situation, the `Event` trait has no way to communicate that its default shouldn't be used: the `default_value_for` method succeeds, and indicates that the default value is a constant, and that that constant value `Undefined`. That means that a trait like `Tuple(Event(), Event())` has a default value of `(Undefined, Undefined)`. We could potentially update the logic in traits to check for `Undefined` everywhere and interpret that as "no default", but it seems cleaner to introduce a new default value type for this.

This PR doesn't fix #1541, but it should enable an easy fix for that issue.

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`)
- ~[ ] Update User manual (`docs/source/traits_user_manual`)~ N/A
- [x] Update type annotation hints in `traits-stubs`

---

Update: we renamed from `DefaultValue.unsupported` to `DefaultValue.disallow`; I've edited the description and title accordingly.

### CODE REVIEW

File: traits/constants.py [hunk 1]
```diff
@@ -215,4 +219,5 @@ class DefaultValue(IntEnum):
     DefaultValue.callable_and_args: "factory",
     DefaultValue.callable: "method",
     DefaultValue.trait_set_object: "set",
+    DefaultValue.unsupported: "unsupported"
```
Comment (mdickinson) [inline traits/constants.py:222]:
```suggestion
    DefaultValue.unsupported: "unsupported",
```

Comment (mdickinson) [inline traits/constants.py]: The `default_kind` property is essentially unused; I'm including this only for completeness.

**Comment by mdickinson** [comment]
There's also potential to use this in the future for `Instance` (or other) traits where we want to enforce that the default of `None` is never used in practice.

**Comment by mdickinson** [comment]
I'm not wedded to the "unsupported" name here; alternative naming suggestions welcome. I was also considering "disallow", to match the `Disallow` trait type.

**Review by rahulporuri — COMMENTED** [review]
IIRC from the internal discussion, we decided to rename this to `disallowed` instead of `unsupported`.

**Comment by mdickinson** [comment]
> IIRC from the internal discussion, we decided to rename this to `disallowed` instead of `unsupported`.

Yes, I'll rename (but I think to "disallow" rather than "disallowed").

**Comment by mdickinson** [comment]
Updated for the suggested renaming; I've also edited the PR title and description to match to make it easier to find in the future.

File: traits-stubs/traits-stubs/constants.pyi [hunk 2]
```diff
@@ -64,6 +64,8 @@ class DefaultValue(IntEnum):
     callable_and_args: int = ...
     callable: int = ...
     trait_set_object: int = ...
+    unsupported: int = ...
```
Comment (mdickinson) [inline traits-stubs/traits-stubs/constants.pyi:67]:
```suggestion
    disallow: int = ...
```

Comment (mdickinson) [reply traits-stubs/traits-stubs/constants.pyi]: Sorry, I missed this one. Fixed now.

File: traits/ctraits.c [hunk 3]
```diff
@@ -1845,6 +1848,12 @@ default_value_for(trait_object *trait, has_traits_object *obj, PyObject *name)
         case TRAIT_SET_OBJECT_DEFAULT_VALUE:
             return call_class(
                 TraitSetObject, trait, obj, name, trait->default_value);
+        case DISALLOW_DEFAULT_VALUE:
+            PyErr_SetString(
+                PyExc_ValueError,
+                "default value unsupported for this trait"
```
Comment (mdickinson) [inline traits/ctraits.c:1854]:
```suggestion
                "default value not permitted for this trait"
```

**Review by corranwebster — APPROVED** [review]

File: traits/constants.py [hunk 4]
```diff
@@ -215,4 +219,5 @@ class DefaultValue(IntEnum):
     DefaultValue.callable_and_args: "factory",
     DefaultValue.callable: "method",
     DefaultValue.trait_set_object: "set",
+    DefaultValue.disallow: "invalid",
```
Comment (rahulporuri) [inline traits/constants.py:222]:
```suggestion
    DefaultValue.disallow: "disallow",
```

File: traits/tests/test_trait_types.py [hunk 5]
```diff
@@ -110,6 +110,28 @@ class MyTraitType(TraitType):
...
+            trait_type.get_default_value(),
+            (DefaultValue.disallow, Undefined)
+        )
+
+        ctrait = trait_type.as_ctrait()
+        self.assertEqual(
+            ctrait.default_value(),
+            (DefaultValue.disallow, Undefined),
+        )
+        self.assertEqual(ctrait.default_kind, "invalid")
```
Comment (rahulporuri) [inline traits/tests/test_trait_types.py:129]:
```suggestion
        self.assertEqual(ctrait.default_kind, "disallow")
```

File: traits/constants.py (hunk 4 above)
Comment (mdickinson) [reply traits/constants.py:222]: Yep, the choice not to use "disallow" here was deliberate. In as much as the `default_kind` is used for anything, it's a description of the type of the default value (`"value"`, `"set"`, `"method"`, etc.), and `"invalid"` seemed a better fit than `"disallow"` here.

Comment (mdickinson) [reply traits/constants.py:222]: (Bigger picture: one of my goals with all this refactoring is to make the default value handling consistent and clean enough that the `default_kind` property is clearly not useful.)

Comment (mdickinson) [reply traits/constants.py:222]: We could always rename from "disallow" to "invalid" throughout, if anyone feels it's important enough.

Comment (rahulporuri) [reply traits/constants.py:222]: ahh. sounds good to me. resolving.

**Pull request merged**